flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDataFileSystemFormatFactory.java

```diff
@@ -0,0 +1,352 @@
+		private final String[] fieldNames;
+		private final int[] selectFields;
+		private final List<String> partitionKeys;
+		private final String defaultPartValue;
+		private final long limit;
+		private final int[] csvSelectFieldToProjectFieldMapping;
+		private final int[] csvSelectFieldToCsvFieldMapping;
+		private final boolean ignoreParseErrors;
+
+		private transient InputStreamReader inputStreamReader;
+		private transient BufferedReader reader;
+		private transient boolean end;
+		private transient long emitted;
+		// reuse object for per record
+		private transient GenericRowData rowData;
+		private transient DeserializationRuntimeConverter runtimeConverter;
+		private transient MappingIterator<JsonNode> iterator;
+
+		public CsvInputFormat(
+				Path[] filePaths,
+				DataType[] fieldTypes,
+				String[] fieldNames,
+				CsvSchema csvSchema,
+				RowType formatRowType,
+				int[] selectFields,
+				List<String> partitionKeys,
+				String defaultPartValue,
+				long limit,
+				int[] csvSelectFieldToProjectFieldMapping,
+				int[] csvSelectFieldToCsvFieldMapping,
+				boolean ignoreParseErrors) {
+			super(filePaths, csvSchema);
+			this.fieldTypes = fieldTypes;
+			this.fieldNames = fieldNames;
+			this.formatRowType = formatRowType;
+			this.partitionKeys = partitionKeys;
+			this.defaultPartValue = defaultPartValue;
+			this.selectFields = selectFields;
+			this.limit = limit;
+			this.emitted = 0;
+			this.csvSelectFieldToProjectFieldMapping = csvSelectFieldToProjectFieldMapping;
+			this.csvSelectFieldToCsvFieldMapping = csvSelectFieldToCsvFieldMapping;
+			this.ignoreParseErrors = ignoreParseErrors;
+		}
+
+		@Override
+		public void open(FileInputSplit split) throws IOException {
+			super.open(split);
+			this.end = false;
+			this.inputStreamReader = new InputStreamReader(csvInputStream);
+			this.reader = new BufferedReader(inputStreamReader);
+			this.rowData = PartitionPathUtils.fillPartitionValueForRecord(fieldNames, fieldTypes, selectFields,
+				partitionKeys, currentSplit.getPath(), defaultPartValue);
+			this.iterator = new CsvMapper()
+				.readerFor(JsonNode.class)
+				.with(csvSchema)
+				.readValues(csvInputStream);
+			prepareRuntimeConverter();
+		}
+
+		private void prepareRuntimeConverter(){
+			CsvRowDataDeserializationSchema.Builder builder = new CsvRowDataDeserializationSchema.Builder(
+				formatRowType, new GenericTypeInfo<>(RowData.class))
+				.setIgnoreParseErrors(ignoreParseErrors);
+			this.runtimeConverter = builder.build().createRowConverter(formatRowType, true);
+		}
+
+		@Override
+		public boolean reachedEnd() throws IOException {
+			return emitted >= limit || end;
+		}
+
+		@Override
+		public RowData nextRecord(RowData reuse) throws IOException {
+			GenericRowData csvRow = null;
+			do {
+				try {
+					JsonNode root = iterator.nextValue();
+					csvRow = (GenericRowData) runtimeConverter.convert(root);
+				} catch (NoSuchElementException e) {
+					end = true;
```

Comment:
`return null` after `end = true`. And remove `!reachedEnd()` in `while`.
This can make codes clear and correct `emitted` number.

Reply:
I update the logic, please review again.